tools/Common.psm1

```diff
@@ -0,0 +1,553 @@
+        {
+            throw "Failed to assign role Owner to service principal $($servicePrincipal.ClientId)"
+        }
+        $loginInfo["ClientId"] = $servicePrincipal.ClientId
+        # The client secret in ADFS is in plain text, no need to convert from SecureString for now. Reconsider this in the future if this changes.
+        $loginInfo["ClientSecret"] = $servicePrincipal.ClientSecret | ConvertTo-SecureString -AsPlainText -Force | ConvertFrom-SecureString
+        $loginInfo["ObjectId"] = $servicePrincipal.ApplicationIdentifier
+            
+        # Clean up and reset context.
+        $session | Remove-PSSession
+    }
+    else
+    {
+        Write-Verbose -Message "Creating Service Principal ${ServicePrincipalName}." -Verbose
+        $servicePrincipal = Get-AzADServicePrincipal -DisplayName $ServicePrincipalName
+
+        if ($servicePrincipal)
+        {
+            Remove-AzADServicePrincipal -ObjectId $servicePrincipal.Id -Force
+        }
+
+        $servicePrincipalApplication = Get-AzADApplication -DisplayName $ServicePrincipalName
+
+        if ($servicePrincipalApplication)
+        {
+            $servicePrincipalApplication | Remove-AzADApplication -Force
+        }
+
+        $servicePrincipal = New-AzADServicePrincipal -Role Owner -DisplayName $ServicePrincipalName
+        $spExistsExpression = "((Get-AzADServicePrincipal -ObjectId $($servicePrincipal.Id)) -and (Get-AzADApplication -DisplayName ${ServicePrincipalName}))"
+        $spExists = Wait-IncrementsUntilTimeOut -CheckExpression $spExistsExpression -Negate
+        if (!$spExists)
+        {
+            throw "The service principal $($servicePrincipal.Id) still failed to exist after checking for an alloted amount of time."
+        }
+
+        Write-Verbose -Message "Created the following service principal:" -Verbose
+        $servicePrincipal | Out-String | Write-Verbose -Verbose
+        
+        $loginInfo["ClientId"] = $servicePrincipal.ApplicationId
+        $loginInfo["ClientSecret"] = ConvertFrom-SecureString $servicePrincipal.Secret
+        $loginInfo["ObjectId"] = $servicePrincipal.Id
+    }
+
+    if (-not $servicePrincipal)
+    {
+        throw [System.Exception] "The service principal is null. Failed to create and connect to a service principal context named ${ServicePrincipalName}."
+    }
+
+    return $loginInfo
+}
+
+function New-AzureCertSp
+{
+    [CmdletBinding(DefaultParameterSetName = "notAdfs")]
+    param
+    (
+        [Parameter(ParameterSetName = "notAdfs", Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ServicePrincipalName,
+
+        [Parameter(ParameterSetName = "Adfs", Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ApplicationName,
+
+        [Parameter(ParameterSetName = "notAdfs", Mandatory = $false)]
+        [Parameter(ParameterSetName = "Adfs", Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ErcsVmDnsName= "AzS-ERCS01",
```

Comment:
can take from BVTSettings, and set correctly for 1 node and 4 node.

Reply:
Duplicate comment.